python/lsst/ip/diffim/imageMapReduce.py

```diff
@@ -0,0 +1,575 @@
+
+        Split the exposure into sub-exps on a grid (parameters given
+        by `ImageMapReduceConfig`) and perform `mapperSubtask.run()` on
+        each. Reduce the resulting sub-exps by running `reducerSubtask.run()`.
+
+        Parameters
+        ----------
+        exposure : afwImage.ExposureF
+            the full exposure to process
+        kwargs :
+            additional keyword arguments to be passed to
+            subtask `run` methods
+
+        Returns
+        -------
+        output of `reducerSubtask.run()`
+        """
+        self.log.info("Processing.")
+
+        patches = self._runMapper(exposure, **kwargs)
+        result = self._reduceImage(patches, exposure, **kwargs)
+        return result
+
+    def _runMapper(self, exposure, doClone=False, **kwargs):
+        """Perform `mapperSubtask.run()` on each patch
+
+        Perform `mapperSubtask.run()` on each patch across a
+        grid on `exposure` generated by `generateGrid`.
+
+        Parameters
+        ----------
+        exposure : afwImage.ExposureF
+            the original exposure which is used as the template
+        doClone : boolean
+            if True, clone the subimages before passing to subtask;
+            in that case, the sub-exps do not have to be considered as read-only
+        kwargs :
+            additional keyword arguments to be passed to
+            `mapperSubtask.run`
+
+        Returns
+        -------
+        a list of `afwExposure`s or other values returned by `mapperSubtask.run`.
+        """
+        boxes0, boxes1 = self.boxes0, self.boxes1
+        if boxes0 is None:
+            boxes0, boxes1 = self._generateGrid(exposure)
+        if len(boxes0) != len(boxes1):
+            raise Exception('Uh oh!')   # TBD: define a specific exception to raise
+
+        self.log.info("Processing %d sub-exposures" % len(boxes0))
+        patches = []
+        for i in range(len(boxes0)):
+            subExp = afwImage.ExposureF(exposure, boxes0[i])
+            expandedSubExp = afwImage.ExposureF(exposure, boxes1[i])
+            if doClone:
+                subExp = subExp.clone()
+                expandedSubExp = expandedSubExp.clone()
+            result = self.mapperSubtask.run(subExp, expandedSubExp, exposure.getBBox(), **kwargs)
+            patches.append(result)
+
+        return patches
+
+    def _reduceImage(self, patches, exposure, **kwargs):
+        """Reduce/merge a set of sub-exposure patches into a final exposure
+
+        Return an exposure of the same dimensions as `exposure`.
+        `patches` is expected to have been produced by `runMapper`.
+
+        Parameters
+        ----------
+        patches : list
+            list of subExposures of `exposure` to merge
+        exposure : afwImage.Exposure
+            the original exposure
+        **kwargs :
+            additional keyword arguments
+
+        Returns
+        -------
+        Output of `reducerSubtask.run` which may be an `afw.Exposure` or
```

Comment:
Upon further thought, I'm not sure if pipeBase.Struct is the right thing here. In all cases, if I were to use `pipeBase.Struct` it would contain a single item -- either an exposure or a list. I suppose I could enforce that it either return a `pipeBase.Struct(exposure=result)` or `pipeBase.Struct(items=result)`. Thoughts?